### PR TITLE
Fix mypy typing errors in pytorch_lightning/callbacks/model_checkpoint.py

### PR DESCRIPTION
**Pull request by BongYang**
## What does this PR do?
Fix typing errors in pytorch_lightning/callbacks/model_checkpoint.py in issue https://github.com/Lightning-AI/lightning/issues/13445.

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
LGTM! I pushed some changes fixing some of the types

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -608,12 +610,12 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
 
         self.dirpath = ckpt_path
 
-    def __warn_if_dir_not_empty(self, dirpath: _PATH) -> None:
-        if self.save_top_k != 0 and self._fs.isdir(dirpath) and len(self._fs.ls(dirpath)) > 0:
+    def __warn_if_dir_not_empty(self, dirpath: Optional[_PATH]) -> None:
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:613]: But dirpath can't be None, right? This type will never be satisfied. 
After the call `self.__resolve_ckpt_dir(trainer)` the dirpath is guaranteed to be not None.

File: src/pytorch_lightning/strategies/dp.py [hunk 2]
```diff
@@ -125,8 +125,8 @@ def barrier(self, *args: Any, **kwargs: Any) -> None:
     def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
         return obj
 
-    def reduce_boolean_decision(self, decision: bool) -> bool:
-        return decision
+    def reduce_boolean_decision(self, decision: Union[bool, Tensor]) -> bool:
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/dp.py:128]: I disagree with the change. Let's instead convert to bool where necessary when calling the method. I don't see a reason why we would want to allow a tensor.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: src/pytorch_lightning/strategies/dp.py (hunk 2 above)
Comment (carmocca) [reply src/pytorch_lightning/strategies/dp.py:128]: In the current `ModelCheckpoint` implementation, `reduce_boolean_decision` is called with a Tensor input: https://github.com/Lightning-AI/lightning/blob/35ec79ad1f75790767a6857637c9bbe052e47b58/src/pytorch_lightning/callbacks/model_checkpoint.py#L496

The current implementation in master supports this, but it's inefficient. It converts the tensor to int and then back to a tensor: https://github.com/Lightning-AI/lightning/blob/35ec79ad1f75790767a6857637c9bbe052e47b58/src/pytorch_lightning/strategies/parallel.py#L116

So master already allows a Tensor, the changes in this PR just fix the type and also makes the implementation a bit better by not creating a new tensor type.

@awaelchli do you still want to manually convert to `bool` in `ModelCheckpoint` before calling `reduce_boolean_decision`?

Comment (awaelchli) [reply src/pytorch_lightning/strategies/dp.py:128]: Yes I prefer to keep it to bool. Otherwise one has to deal with non-scalar tensors and that just makes it unnecessarily complicated.

Comment (carmocca) [reply src/pytorch_lightning/strategies/dp.py]: Done

File: src/pytorch_lightning/callbacks/early_stopping.py [hunk 3]
```diff
@@ -135,7 +135,7 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: O
             # validation, then we run after validation instead of on train epoch end
             self._check_on_train_epoch_end = trainer.val_check_interval == 1.0 and trainer.check_val_every_n_epoch == 1
 
-    def _validate_condition_metric(self, logs: Dict[str, float]) -> bool:
+    def _validate_condition_metric(self, logs: Dict[str, Tensor]) -> bool:
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/early_stopping.py:138]: `logs` should be the return type of `logger_connector.callback_metrics`

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -362,7 +364,7 @@ def save_checkpoint(self, trainer: "pl.Trainer") -> None:  # pragma: no-cover
         self._save_topk_checkpoint(trainer, monitor_candidates)
         self._save_last_checkpoint(trainer, monitor_candidates)
 
-    def _save_topk_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
+    def _save_topk_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, Tensor]) -> None:
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:367]: same here

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -395,7 +397,7 @@ def _should_skip_saving_checkpoint(self, trainer: "pl.Trainer") -> bool:
         from pytorch_lightning.trainer.states import TrainerFn
 
         return (
-            trainer.fast_dev_run  # disable checkpointing with fast_dev_run
+            bool(trainer.fast_dev_run)  # disable checkpointing with fast_dev_run
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:400]: why bool?

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -493,15 +495,15 @@ def check_monitor_top_k(self, trainer: "pl.Trainer", current: Optional[Tensor] =
         should_update_best_and_save = monitor_op(current, self.best_k_models[self.kth_best_model_path])
 
         # If using multiple devices, make sure all processes are unanimous on the decision.
-        should_update_best_and_save = trainer.strategy.reduce_boolean_decision(should_update_best_and_save)
+        should_update_best_and_save = trainer.strategy.reduce_boolean_decision(bool(should_update_best_and_save))
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:498]: same here

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -531,7 +533,7 @@ def _format_checkpoint_name(
         return filename
 
     def format_checkpoint_name(
-        self, metrics: Dict[str, _METRIC], filename: Optional[str] = None, ver: Optional[int] = None
+        self, metrics: Dict[str, Tensor], filename: Optional[str] = None, ver: Optional[int] = None
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:536]: same here from `logger_connector.callback_metrics`.

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -613,7 +616,7 @@ def __warn_if_dir_not_empty(self, dirpath: _PATH) -> None:
             rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
 
     def _get_metric_interpolated_filepath_name(
-        self, monitor_candidates: Dict[str, _METRIC], trainer: "pl.Trainer", del_filepath: Optional[str] = None
+        self, monitor_candidates: Dict[str, Tensor], trainer: "pl.Trainer", del_filepath: Optional[str] = None
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:619]: same here from `logger_connector.callback_metrics`.

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 9]
```diff
@@ -624,7 +627,7 @@ def _get_metric_interpolated_filepath_name(
 
         return filepath
 
-    def _monitor_candidates(self, trainer: "pl.Trainer") -> Dict[str, _METRIC]:
+    def _monitor_candidates(self, trainer: "pl.Trainer") -> Dict[str, Tensor]:
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:630]: same here from `logger_connector.callback_metrics`.

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 10]
```diff
@@ -634,7 +637,7 @@ def _monitor_candidates(self, trainer: "pl.Trainer") -> Dict[str, _METRIC]:
         monitor_candidates["step"] = step.int() if isinstance(step, Tensor) else torch.tensor(trainer.global_step)
         return monitor_candidates
 
-    def _save_last_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
+    def _save_last_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, Tensor]) -> None:
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:640]: same here from `logger_connector.callback_metrics`.

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 11]
```diff
@@ -651,16 +654,18 @@ def _save_last_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[
         if previous and previous != filepath:
             trainer.strategy.remove_checkpoint(previous)
 
-    def _save_monitor_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
+    def _save_monitor_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, Tensor]) -> None:
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:657]: same here from `logger_connector.callback_metrics`.

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 5 above)
Comment (carmocca) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:400]: mypy seems to be strict in this regard. It doesn't like that we check it like this if it can also be an int

File: src/pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (carmocca) [reply src/pytorch_lightning/callbacks/early_stopping.py:138]: I considered doing this. But didn't want to in this PR

The `callback_metrics` annotation is internal. I fear this creates circular imports and perhaps we should make it public and move it to `types.py`

I think this should be done in a follow-up

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 6 above)
Comment (carmocca) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:498]: See discussion here: https://github.com/Lightning-AI/lightning/pull/13617#discussion_r921182898

File: src/pytorch_lightning/trainer/trainer.py [hunk 12]
```diff
@@ -2705,7 +2705,9 @@ def loggers(self, loggers: Optional[List[Logger]]) -> None:
         self._loggers = loggers if loggers else []
 
     @property
-    def callback_metrics(self) -> dict:
+    def callback_metrics(self) -> Dict[str, Tensor]:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:2708]: Answered in https://github.com/Lightning-AI/lightning/pull/13617/files#r924445862

**Review by awaelchli — APPROVED** [review]
